File: scripts/ops/sources.py
"""Bounded GET-only collectors, with explicit coverage for every source.

Raw responses stay internal. Renderers receive only the evidence allowlist.
"""
from concurrent.futures import ThreadPoolExecutor
from datetime import datetime, timezone
import json
import os
import re
import subprocess
from urllib.error import HTTPError, URLError
from urllib.parse import urlencode, quote
from urllib.request import Request, build_opener, HTTPRedirectHandler

from .config import Config, ROOT, WORKFLOWS


class SourceError(Exception):
    """Safe to display: never contains subprocess stderr or HTTP response bodies."""


def github_get(path: str):
    try:
        result = subprocess.run(["gh", "api", "--method", "GET", path],
                                capture_output=True, text=True, timeout=30,
                                env={**os.environ, "GH_HOST": "github.com", "GH_PROMPT_DISABLED": "1"})
        if result.returncode:
            raise SourceError("GitHub request failed; check gh auth, read permissions and rate limits")
        return json.loads(result.stdout)
    except (OSError, subprocess.TimeoutExpired, ValueError):
        raise SourceError("GitHub unavailable; install/authenticate gh or check connection") from None


class NoRedirect(HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None


def vercel_get(path: str):
    token = os.environ.get("VERCEL_TOKEN")
    if not token:
        raise SourceError("VERCEL_TOKEN not set")
    try:
        req = Request("https://api.vercel.com" + path, headers={"Authorization": f"Bearer {token}"}, method="GET")
        with build_opener(NoRedirect()).open(req, timeout=30) as response:
            payload = response.read(16 * 1024 * 1024 + 1)
            if len(payload) > 16 * 1024 * 1024:
                raise SourceError("Vercel response exceeded size limit")
            return json.loads(payload)
    except HTTPError as error:
        raise SourceError(f"Vercel HTTP {error.code}; check token, scope, project ID or rate limits") from None
    except (URLError, OSError, ValueError):
        raise SourceError("Vercel unavailable or invalid response") from None


def full_sha(value):
    return value.lower() if isinstance(value, str) and re.fullmatch(r"[a-fA-F0-9]{40}", value) else None


def tag_time(ref: str) -> str:
    match = re.search(r"/(\d{4}-\d\d-\d\dT\d\d)-(\d\d)-(\d\dZ)/", ref)
    return f"{match[1]}:{match[2]}:{match[3]}" if match else ""


class Collector:
    def __init__(self, config: Config, limit: int, pages: int, gh=github_get, vercel=vercel_get):
        self.config, self.limit, self.pages = config, limit, pages
        self.gh, self.vercel = gh, vercel
        self.base = f"repos/{config.repo}"
        self.coverage = []

    def note(self, source: str, state: str, detail: str):
        self.coverage.append({"source": source, "state": state, "detail": detail})

    def get(self, path: str, label: str):
        try:
            return self.gh(f"{self.base}/{path}")
        except SourceError as error:
            self.note(label, "unavailable", str(error))
            return None

    def listing(self, path: str, label: str, key: str | None = None, page_size=100, pages=None):
        records = []
        seen = set()
        for page in range(1, (pages or self.pages) + 1):
            sep = "&" if "?" in path else "?"
            data = self.get(f"{path}{sep}per_page={page_size}&page={page}", label)
            if data is None:
                return records
            items = data.get(key) if key and isinstance(data, dict) else data
            if not isinstance(items, list):
                self.note(label, "unavailable", "Unexpected API response")
                return records
            fingerprint = json.dumps(items, sort_keys=True)
            if items and fingerprint in seen:
                self.note(label, "partial", "Pagination repeated a page; stopped")
                return records
            seen.add(fingerprint)
            records.extend(items)
            total = data.get("total_count") if isinstance(data, dict) else None
            if len(items) < page_size or (isinstance(total, int) and len(records) >= total):
                self.note(label, "complete", f"{len(records)} records")
                return records
        self.note(label, "windowed", f"{len(records)} records; page cap reached (increase --pages / --limit)")
        return records

    @staticmethod
    def parallel(function, items):
        with ThreadPoolExecutor(max_workers=6) as pool:
            return list(pool.map(function, items))

    def tags(self, prefix: str, sha: str | None):
        # Git matching-refs has no pagination parameters; unlike /tags it returns
        # refs (including the annotated-tag object ID), not peeled commits.
        refs = self.get(f"git/matching-refs/tags/{prefix}", f"{prefix} refs")
        if not isinstance(refs, list):
            return []
        self.note(f"{prefix} refs", "complete", f"{len(refs)} refs")
        if prefix == "deploy/":
            refs.sort(key=lambda r: (tag_time(r["ref"]), r["ref"]), reverse=True)
        else:
            refs = [r for r in refs if re.fullmatch(r"refs/tags/v\d+\.\d+\.\d+", r["ref"])]
            refs.sort(key=lambda r: tuple(map(int, r["ref"].split("/v")[-1].split("."))), reverse=True)
        if prefix == "deploy/":
            selected = [r for env in ("staging", "production")
                        for r in [r for r in refs if r["ref"].startswith(f"refs/tags/deploy/{env}/")][:self.limit]]
        else:
            selected = refs[:self.limit]
        if sha:
            selected += [r for r in refs if r["ref"].endswith("/" + sha) and r not in selected]
        if len(selected) < len(refs):
            self.note(f"{prefix} tag details", "windowed", f"Resolved {len(selected)} of {len(refs)} refs")

        def resolve(ref):
            obj, message = ref.get("object", {}), ""
            for _ in range(5):
                if obj.get("type") != "tag":
                    break
                detail = self.get(f"git/tags/{obj['sha']}", "tag detail")
                if not detail:
                    return None
                message += "\n" + detail.get("message", "")
                obj = detail.get("object", {})
            target = full_sha(obj.get("sha")) if obj.get("type") == "commit" else None
            if not target:
                self.note("tag detail", "partial", "Tag does not resolve to a full commit SHA")
                return None
            name = ref["ref"].removeprefix("refs/tags/")
            # A suffix alone is never sufficient to establish a deployed commit.
            if prefix == "deploy/" and name.rsplit("/", 1)[-1] != target:
                self.note("tag detail", "partial", "Deployment tag suffix and target differ; ignored")
                return None
            match = re.search(r"Workflow run: https://github\.com/" + re.escape(self.config.repo) + r"/actions/runs/(\d+)\b", message)
            return {"name": name, "sha": target, "time": tag_time(ref["ref"]),
                    "run_id": int(match[1]) if match else None}
        return [t for t in self.parallel(resolve, selected) if t]

    def collect(self, sha: str | None = None, no_vercel=False):
        tags = self.tags("deploy/", sha) + self.tags("v", None)
        workflows = list(WORKFLOWS) + [f"ci-{app}.yml" for app in self.config.apps
                                       if (ROOT / ".github/workflows" / f"ci-{app}.yml").is_file()]
        def workflow_runs(workflow):
            query = f"&head_sha={sha}" if sha and workflow not in ("cd-production.yml", "cd-rollback.yml") else ""
            return self.listing(f"actions/workflows/{workflow}/runs?exclude_pull_requests=true{query}",
                                workflow, "workflow_runs", self.limit, 1)
        runs = [r for batch in self.parallel(workflow_runs, workflows) for r in batch]
        # Fetch tagged runs even if older than the run window. Their annotations
        # establish the requested SHA of production / rollback dispatches.
        ids = {r["id"] for r in runs}
        for tag in tags:
            run_id = tag["run_id"]
            if run_id and run_id not in ids:
                run = self.get(f"actions/runs/{run_id}", "tag workflow")
                if run:
                    runs.append(run)
                    ids.add(run_id)

        def jobs(run):
            items = self.listing(f"actions/runs/{run['id']}/jobs?filter=latest", f"jobs/{run['id']}", "jobs")
            return [{**job, "run_id": run["id"]} for job in items]
        job_rows = [j for batch in self.parallel(jobs, runs) for j in batch]
        artifacts = self.listing("actions/artifacts", "artifacts", "artifacts")
        # A busy monorepo's test reports can crowd build uploads out of the
        # repository inventory. Inspect every selected CD run as well.
        def run_artifacts(run):
            return self.listing(f"actions/runs/{run['id']}/artifacts", f"artifacts/{run['id']}", "artifacts")
        cd_runs = [r for r in runs if r.get("path", "").split("/")[-1] in WORKFLOWS]
        artifacts += [a for batch in self.parallel(run_artifacts, cd_runs) for a in batch]
        artifacts = list({a["id"]: a for a in artifacts}.values())
        # Recover builder context outside the run window, but only for bundle
        # names this project recognizes (not thousands of test reports).
        app_pattern = "|".join(re.escape(app) for app in self.config.apps)
        builder_ids = {(a.get("workflow_run") or {}).get("id") for a in artifacts
                       if re.fullmatch(rf"(?:{app_pattern})-(?:(?:staging|production|preview)-)?(?:[0-9a-f]{{16}}|[0-9a-f]{{40}})", a.get("name", ""))}
        for run_id in sorted(builder_ids - ids - {None}):
            run = self.get(f"actions/runs/{run_id}", "artifact builder")
            if run:
                runs.append(run)
        deployments = []
        for env in ("staging", "production"):
            deployments += self.listing(f"deployments?environment={env}", f"deployments/{env}", page_size=min(100, self.limit * 4), pages=1)

        def with_status(dep):
            statuses = self.listing(f"deployments/{dep['id']}/statuses", f"deployment status/{dep['id']}")
            return {**dep, "statuses": statuses}
        deployments = self.parallel(with_status, deployments)
        vercel = self.collect_vercel(no_vercel)
        shas = {t["sha"] for t in tags}
        shas.update(full_sha(r.get("head_sha")) for r in runs if r.get("path", "").split("/")[-1] == "cd-staging.yml")
        if sha:
            shas.add(sha)
        shas.discard(None)

        def gate(commit):
            # Combined status returns latest status for each context, not a
            # stale successful status from an earlier workflow attempt.
            data = self.get(f"commits/{commit}/status?per_page=100", f"CI gate/{commit}")
            if data is None:
                return commit, {"state": "unknown"}
            gate = next((s for s in data.get("statuses", []) if s.get("context") == "ci/gate-passed"), None)
            if gate:
                return commit, {k: gate.get(k) for k in ("state", "target_url", "updated_at")}
            return commit, {"state": "missing" if data.get("total_count", 0) <= 100 else "unknown"}
        gates = dict(self.parallel(gate, sorted(shas)))
        return {"repo": self.config.repo, "gathered_at": datetime.now(timezone.utc).isoformat(),
                "tags": tags, "runs": runs, "jobs": job_rows, "artifacts": artifacts,
                "deployments": deployments, "vercel": vercel, "gates": gates,
                "coverage": sorted(self.coverage, key=lambda c: (c["source"], c["state"], c["detail"]))}

    def collect_vercel(self, disabled: bool):
        rows = []
        for app in self.config.apps:
            for env in ("staging", "production"):
                label = f"Vercel/{app}/{env}"
                project = self.config.projects.get(app, {}).get(env)
                if disabled or not project:
                    self.note(label, "unavailable", "Disabled by --no-vercel" if disabled else "Project ID not configured")
                    continue
                params = {"teamId": self.config.team_id} if self.config.team_id else {}
                current = None
                try:
                    detail = self.vercel(f"/v9/projects/{quote(project)}?{urlencode(params)}")
                    target = (detail.get("targets") or {}).get("production") or {}
                    current = target.get("id")
                    if not current:
                        self.note(label + "/current", "partial", "No production target reported; serving deployment unknown")
                except SourceError as error:
                    self.note(label + "/current", "unavailable", str(error))
                except (AttributeError, TypeError):
                    self.note(label + "/current", "partial", "Unexpected Vercel project response")
                try:
                    params.update({"projectId": project, "limit": 100})
                    seen = set()
                    for _ in range(self.pages):
                        data = self.vercel("/v7/deployments?" + urlencode(params))
                        for dep in data["deployments"]:
                            rows.append({"app": app, "environment": env, "project": project,
                                         "deployment": dep, "current": dep.get("uid") == current if current else None})
                        cursor = (data.get("pagination") or {}).get("next")
                        if not cursor:
                            self.note(label, "complete", "Deployment list exhausted")
                            break
                        if cursor in seen:
                            self.note(label, "partial", "Repeated pagination cursor; stopped")
                            break
                        seen.add(cursor)
                        params["until"] = cursor
                    else:
                        self.note(label, "windowed", "Page cap reached; increase --pages")
                except SourceError as error:
                    self.note(label, "unavailable", str(error))
                except (KeyError, TypeError, AttributeError):
                    self.note(label, "partial", "Unexpected Vercel response")
                # Preserve project evidence even when listing failed or the
                # serving deployment is older than the history window.
                if current and not any(r["deployment"].get("uid") == current for r in rows):
                    rows.append({"app": app, "environment": env, "project": project,
                                 "deployment": {**target, "uid": current}, "current": True})
        return rows

File: scripts/ops/sources.ts
/** Bounded GET-only collectors. Raw responses and subprocess stderr never reach output. */
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import { existsSync } from "node:fs";
import { resolve } from "node:path";
import { ENVIRONMENTS, ROOT, WORKFLOWS } from "./config.js";
import { errorMessage, fullSha, object } from "./types.js";
import { annotationRunId } from "./urls.js";
import type {
  Artifact,
  Config,
  Coverage,
  DeploymentStatus,
  Gate,
  GitHubDeployment,
  Job,
  Ref,
  Run,
  Snapshot,
  Tag,
  TagObject,
  VercelDeployment,
  VercelRow,
} from "./types.js";

const execute = promisify(execFile);
export class SourceError extends Error {}
export type Get = (path: string) => Promise<unknown>;
export type Command = (file: string, args: string[]) => Promise<string>;
export function createGithubGet(
  run: Command = async (file, args) => {
    const { stdout } = await execute(file, args, {
      encoding: "utf8",
      timeout: 30_000,
      maxBuffer: 16 * 1024 * 1024,
      env: { ...process.env, GH_HOST: "github.com", GH_PROMPT_DISABLED: "1" },
    });
    return stdout;
  },
): Get {
  return async (path) => {
    try {
      return JSON.parse(
        await run("gh", ["api", "--method", "GET", path]),
      ) as unknown;
    } catch {
      throw new SourceError(
        "GitHub unavailable; check gh installation/authentication, read permissions, connection and rate limits",
      );
    }
  };
}
export const githubGet = createGithubGet();
export const vercelGet: Get = async (path) => {
  const token = process.env.VERCEL_TOKEN;
  if (!token) throw new SourceError("VERCEL_TOKEN not set");
  try {
    const response = await fetch(`https://api.vercel.com${path}`, {
      method: "GET",
      headers: { Authorization: `Bearer ${token}` },
      redirect: "error",
      signal: globalThis.AbortSignal.timeout(30_000),
    });
    if (!response.ok)
      throw new SourceError(
        `Vercel HTTP ${response.status}; check token, scope, project ID or rate limits`,
      );
    const reader = response.body?.getReader();
    if (!reader) throw new SourceError("Vercel response has no body");
    const chunks: Uint8Array[] = [];
    let size = 0;
    for (;;) {
      const { value, done } = await reader.read();
      if (done) break;
      size += value.byteLength;
      if (size > 16 * 1024 * 1024) {
        await reader.cancel();
        throw new SourceError("Vercel response exceeded size limit");
      }
      chunks.push(value);
    }
    return JSON.parse(Buffer.concat(chunks).toString("utf8")) as unknown;
  } catch (error) {
    if (error instanceof SourceError) throw error;
    throw new SourceError("Vercel unavailable or invalid response");
  }
};
export function tagTime(ref: string): string {
  const match = /\/(\d{4}-\d\d-\d\dT\d\d)-(\d\d)-(\d\dZ)\//.exec(ref);
  return match ? `${match[1]}:${match[2]}:${match[3]}` : "";
}
const escape = (text: string): string =>
  text.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
export async function parallel<T, U>(
  items: T[],
  fn: (item: T) => Promise<U>,
): Promise<U[]> {
  const results: U[] = new Array(items.length);
  let next = 0;
  await Promise.all(
    Array.from({ length: Math.min(6, items.length) }, async () => {
      while (next < items.length) {
        const index = next++;
        results[index] = await fn(items[index]);
      }
    }),
  );
  return results;
}
export class Collector {
  readonly coverage: Coverage[] = [];
  readonly base: string;
  constructor(
    readonly config: Config,
    readonly limit: number,
    readonly pages: number,
    readonly gh: Get = githubGet,
    readonly vercel: Get = vercelGet,
  ) {
    this.base = `repos/${config.repo}`;
  }
  note(source: string, state: Coverage["state"], detail: string): void {
    this.coverage.push({ source, state, detail });
  }
  async get(path: string, label: string): Promise<unknown> {
    try {
      return await this.gh(`${this.base}/${path}`);
    } catch (error) {
      this.note(
        label,
        "unavailable",
        error instanceof SourceError
          ? error.message
          : "Unexpected GitHub transport failure",
      );
      return null;
    }
  }
  async record<T>(path: string, label: string): Promise<T | null> {
    const data = await this.get(path, label);
    if (data === null) return null;
    if (typeof data !== "object" || Array.isArray(data)) {
      this.note(label, "partial", "Unexpected API response");
      return null;
    }
    return data as T;
  }
  async listing<T>(
    path: string,
    label: string,
    key?: string,
    pageSize = 100,
    pages = this.pages,
  ): Promise<T[]> {
    const records: T[] = [];
    const seen = new Set<string>();
    for (let page = 1; page <= pages; page++) {
      const data = await this.get(
        `${path}${path.includes("?") ? "&" : "?"}per_page=${pageSize}&page=${page}`,
        label,
      );
      if (data === null) return records;
      const items: unknown = key ? object(data)[key] : data;
      if (!Array.isArray(items)) {
        this.note(label, "unavailable", "Unexpected API response");
        return records;
      }
      const fingerprint = JSON.stringify(items);
      if (items.length && seen.has(fingerprint)) {
        this.note(label, "partial", "Pagination repeated a page; stopped");
        return records;
      }
      seen.add(fingerprint);
      records.push(...(items as T[]));
      const total = object(data).total_count;
      if (
        items.length < pageSize ||
        (typeof total === "number" && records.length >= total)
      ) {
        this.note(label, "complete", `${records.length} records`);
        return records;
      }
    }
    this.note(
      label,
      "windowed",
      `${records.length} records; page cap reached (increase --pages / --limit)`,
    );
    return records;
  }
  async tags(prefix: string, sha?: string): Promise<Tag[]> {
    // matching-refs has no pagination parameters; it returns annotated object IDs.
    const data = await this.get(
      `git/matching-refs/tags/${prefix}`,
      `${prefix} refs`,
    );
    if (!Array.isArray(data)) return [];
    let refs = data as Ref[];
    this.note(`${prefix} refs`, "complete", `${refs.length} refs`);
    if (prefix === "deploy/")
      refs.sort(
        (a, b) =>
          tagTime(b.ref).localeCompare(tagTime(a.ref)) ||
          b.ref.localeCompare(a.ref),
      );
    else {
      refs = refs.filter((r) => /^refs\/tags\/v\d+\.\d+\.\d+$/.test(r.ref));
      refs.sort((a, b) => {
        const left = a.ref.split("/v").at(-1)!.split(".").map(Number);
        const right = b.ref.split("/v").at(-1)!.split(".").map(Number);
        return right[0] - left[0] || right[1] - left[1] || right[2] - left[2];
      });
    }
    const selected =
      prefix === "deploy/"
        ? ENVIRONMENTS.flatMap((env) =>
            refs
              .filter((r) => r.ref.startsWith(`refs/tags/deploy/${env}/`))
              .slice(0, this.limit),
          )
        : refs.slice(0, this.limit);
    if (sha)
      selected.push(
        ...refs.filter(
          (r) => r.ref.endsWith(`/${sha}`) && !selected.includes(r),
        ),
      );
    if (selected.length < refs.length)
      this.note(
        `${prefix} tag details`,
        "windowed",
        `Resolved ${selected.length} of ${refs.length} refs`,
      );
    const tags = await parallel(selected, async (ref) => {
      let obj = ref.object;
      let message = "";
      for (let depth = 0; depth < 5 && obj.type === "tag"; depth++) {
        const detail = await this.record<TagObject>(
          `git/tags/${obj.sha}`,
          "tag detail",
        );
        if (!detail) return null;
        message += `\n${detail.message ?? ""}`;
        obj = detail.object;
      }
      const target = obj.type === "commit" ? fullSha(obj.sha) : null;
      if (!target) {
        this.note(
          "tag detail",
          "partial",
          "Tag does not resolve to a full commit SHA",
        );
        return null;
      }
      const name = ref.ref.replace(/^refs\/tags\//, "");
      if (prefix === "deploy/" && name.split("/").at(-1) !== target) {
        this.note(
          "tag detail",
          "partial",
          "Deployment tag suffix and target differ; ignored",
        );
        return null;
      }
      return {
        name,
        sha: target,
        time: tagTime(ref.ref),
        run_id: annotationRunId(message, this.config.repo),
      };
    });
    return tags.filter((tag): tag is Tag => tag !== null);
  }
  async collect(sha?: string, noVercel = false): Promise<Snapshot> {
    const tags = [
      ...(await this.tags("deploy/", sha)),
      ...(await this.tags("v")),
    ];
    const workflows = [
      ...WORKFLOWS,
      ...this.config.apps
        .map((app) => `ci-${app}.yml`)
        .filter((file) => existsSync(resolve(ROOT, ".github/workflows", file))),
    ];
    const runs = (
      await parallel(workflows, (workflow) => {
        const query =
          sha && !["cd-production.yml", "cd-rollback.yml"].includes(workflow)
            ? `&head_sha=${sha}`
            : "";
        return this.listing<Run>(
          `actions/workflows/${workflow}/runs?exclude_pull_requests=true${query}`,
          workflow,
          "workflow_runs",
          this.limit,
          1,
        );
      })
    ).flat();
    const ids = new Set(runs.map((r) => r.id));
    for (const tag of tags)
      if (tag.run_id && !ids.has(tag.run_id)) {
        const run = await this.record<Run>(
          `actions/runs/${tag.run_id}`,
          "tag workflow",
        );
        if (run) {
          runs.push(run);
          ids.add(run.id);
        }
      }
    const jobs = (
      await parallel(runs, async (run) =>
        (
          await this.listing<Job>(
            `actions/runs/${run.id}/jobs?filter=latest`,
            `jobs/${run.id}`,
            "jobs",
          )
        ).map((job) => ({ ...job, run_id: run.id })),
      )
    ).flat();
    const inventory = await this.listing<Artifact>(
      "actions/artifacts",
      "artifacts",
      "artifacts",
    );
    inventory.push(
      ...(
        await parallel(
          runs.filter((r) =>
            WORKFLOWS.includes(r.path?.split("/").at(-1) ?? ""),
          ),
          (run) =>
            this.listing<Artifact>(
              `actions/runs/${run.id}/artifacts`,
              `artifacts/${run.id}`,
              "artifacts",
            ),
        )
      ).flat(),
    );
    const artifacts = [...new Map(inventory.map((a) => [a.id, a])).values()];
    const names = this.config.apps.map(escape).join("|");
    const bundle = new RegExp(
      `^(?:${names})-(?:(?:staging|production|preview)-)?(?:[0-9a-f]{16}|[0-9a-f]{40})$`,
    );
    const builders = new Set(
      artifacts
        .filter((a) => bundle.test(a.name))
        .map((a) => a.workflow_run?.id),
    );
    for (const id of [...builders]
      .filter((id): id is number => id !== undefined && !ids.has(id))
      .sort((a, b) => a - b)) {
      const run = await this.record<Run>(
        `actions/runs/${id}`,
        "artifact builder",
      );
      if (run) runs.push(run);
    }
    const deployments = (
      await parallel(ENVIRONMENTS, (env) =>
        this.listing<GitHubDeployment>(
          `deployments?environment=${env}`,
          `deployments/${env}`,
          undefined,
          Math.min(100, this.limit * 4),
          1,
        ),
      )
    ).flat();
    const withStatuses = await parallel(deployments, async (dep) => ({
      ...dep,
      statuses: await this.listing<DeploymentStatus>(
        `deployments/${dep.id}/statuses`,
        `deployment status/${dep.id}`,
      ),
    }));
    const vercel = await this.collectVercel(noVercel);
    const commits = new Set(tags.map((t) => t.sha));
    for (const run of runs)
      if (run.path?.endsWith("/cd-staging.yml") && fullSha(run.head_sha))
        commits.add(fullSha(run.head_sha)!);
    if (sha) commits.add(sha);
    const gates = Object.fromEntries(
      await parallel([...commits].sort(), async (commit) => {
        const data = await this.record<{
          statuses?: (Gate & { context: string })[];
          total_count?: number;
        }>(`commits/${commit}/status?per_page=100`, `CI gate/${commit}`);
        const gate = data?.statuses?.find(
          (s) => s.context === "ci/gate-passed",
        );
        const result: Gate = gate
          ? {
              state: gate.state,
              target_url: gate.target_url,
              updated_at: gate.updated_at,
            }
          : {
              state: !data
                ? "unknown"
                : (data.total_count ?? 0) <= 100
                  ? "missing"
                  : "unknown",
            };
        return [commit, result];
      }),
    );
    this.coverage.sort(
      (a, b) =>
        a.source.localeCompare(b.source) ||
        a.state.localeCompare(b.state) ||
        a.detail.localeCompare(b.detail),
    );
    return {
      repo: this.config.repo,
      gathered_at: new Date().toISOString(),
      tags,
      runs,
      jobs,
      artifacts,
      deployments: withStatuses,
      vercel,
      gates,
      coverage: this.coverage,
    };
  }
  async collectVercel(disabled: boolean): Promise<VercelRow[]> {
    const rows: VercelRow[] = [];
    for (const app of this.config.apps)
      for (const environment of ENVIRONMENTS) {
        const label = `Vercel/${app}/${environment}`;
        const project = this.config.projects[app]?.[environment];
        if (disabled || !project) {
          this.note(
            label,
            "unavailable",
            disabled ? "Disabled by --no-vercel" : "Project ID not configured",
          );
          continue;
        }
        const params = new URLSearchParams(
          this.config.team_id ? { teamId: this.config.team_id } : {},
        );
        let target: VercelDeployment = {};
        let current: string | undefined;
        try {
          const detail = object(
            await this.vercel(
              `/v9/projects/${encodeURIComponent(project)}?${params}`,
            ),
          );
          target = object(
            object(detail.targets).production,
          ) as VercelDeployment;
          current = target.id;
          if (!current)
            this.note(
              `${label}/current`,
              "partial",
              "No production target reported; serving deployment unknown",
            );
        } catch (error) {
          this.note(
            `${label}/current`,
            "unavailable",
            error instanceof SourceError
              ? error.message
              : "Unexpected Vercel project response",
          );
        }
        try {
          params.set("projectId", project);
          params.set("limit", "100");
          const seen = new Set<string>();
          for (let page = 0; page < this.pages; page++) {
            const data = object(await this.vercel(`/v7/deployments?${params}`));
            if (!Array.isArray(data.deployments))
              throw new SourceError("Unexpected Vercel response");
            for (const deployment of data.deployments as VercelDeployment[])
              rows.push({
                app,
                environment,
                project,
                deployment,
                current: current ? deployment.uid === current : null,
              });
            const cursor = object(data.pagination).next;
            if (cursor === null || cursor === undefined) {
              this.note(label, "complete", "Deployment list exhausted");
              break;
            }
            if (seen.has(String(cursor))) {
              this.note(
                label,
                "partial",
                "Repeated pagination cursor; stopped",
              );
              break;
            }
            seen.add(String(cursor));
            params.set("until", String(cursor));
            if (page + 1 === this.pages)
              this.note(
                label,
                "windowed",
                "Page cap reached; increase --pages",
              );
          }
        } catch (error) {
          this.note(
            label,
            "unavailable",
            error instanceof SourceError
              ? errorMessage(error)
              : "Unexpected Vercel response",
          );
        }
        if (current && !rows.some((r) => r.deployment.uid === current))
          rows.push({
            app,
            environment,
            project,
            deployment: { ...target, uid: current },
            current: true,
          });
      }
    return rows;
  }
}
